pkgs/fc/ceph/src/fc/ceph/lvm.py

```diff
@@ -10,7 +10,92 @@
 from fc.ceph.util import console, mount_status, run
 
 
+class GenericBlockDevice:
+    def __new__(cls, name: str):
+        # prevent explicitly instantiated child classes from returning as None
+        if cls is not GenericLogicalVolume:
+            return object.__new__(cls)
+        if MdraidDevice.exists(name):
+            return MdraidDevice(name)
+        elif PartitionedDisk.exists(name):
+            return PartitionedDisk(name)
+        # Allow operations to proceed by creating a block device
+        return None
+
+    def __init__(self, name: str):
+        self.name = name
+
+    # create cannot be a generic method, as it requires different params depending on the type
+    @property
+    def blockdevice(self) -> str:
+        raise NotImplementedError
+
+    @classmethod
+    def exists(cls, name: str) -> bool:
+        raise NotImplementedError
+
+
+class MdraidDevice(GenericBlockDevice):
+    """represents a software RAID array in mode RAID6 with spare disk. Uses the
+    whole provided blockdevices without partitioning."""
+
+    @classmethod
+    def create(cls, name: str, blockdevices: list[str]):
+        obj = cls(name)
+
+    def __init__(self, name: str):
+        self.name = name
+
+    @property
+    def blockdevice(self) -> str:
+        return f"/dev/md/{self.name}"
+
+    @classmethod
+    def exists(cls, name) -> bool:
+        return name.startswith("/dev/md") and os.path.exists(f"/dev/md/{name}")
+
+    # TODO do we need activating?
+
+
+class PartitionedDisk(GenericBlockDevice):
+    """name denotes the path to the whole unpartitioned disk"""
```

Comment:
rename to `DiskWithSinglePartition`, enhance the docstring that the name denotes the full disk, but the blockdevice is the partition.